### PR TITLE
Allowing failed splits in root search.

### LINKED ISSUE #5411: Be more lenient on missing splits
In the ES API , there is a section to report failed shard.

I am not sure what a good status code should be there. Probably 200 is a safe bet.
206 or 207 are tempting too.

In order to stay on the safe side, we could keep have a strict behavior and a lenient behavior depending on a query parameter.

### PR DESCRIPTION
**Pull request by fulmicoton**
```
    Allowing failed splits in root search.

    This PR adds a query string parameter (`allow_partial_search_results`) in
    the elasticsearch api controlling encounterring an error on a subset
    of the splits, should fail a search query or if a response
    (reflecting the partial set of successful splits) should be returned.

    It changes the behavior of the root search gRPC call, that will not return
    an error in presence of partial split errors, but instead returns the list of splits that failed.

    This PR also changes the default behavior of the elasticsearch rest API: following elasticsearch behavior on failing shards, quickwit will allow split errors.

    The Quickwit API default behavior, on the other hand, is unchanged.

    Closes #5411
```

(I also tested the option by removing splits on a running quickwit instance)

### CODE REVIEW

File: quickwit/quickwit-search/src/collector.rs [hunk 1]
```diff
@@ -1916,6 +1931,7 @@ mod tests {
                     retryable_error: true,
                 }],
                 num_attempted_splits: 5,
+                num_successful_splits: 6,
```
Comment (trinity-1686a) [inline quickwit/quickwit-search/src/collector.rs:1934]: i don't understand, shouldn't that be 3+1 = 4 ?

**Review by trinity-1686a — CHANGES_REQUESTED** [review]

File: quickwit/quickwit-serve/src/elasticsearch_api/rest_handler.rs [hunk 2]
```diff
@@ -853,7 +858,7 @@ async fn es_scroll(
     let search_response: SearchResponse = search_service.scroll(scroll_request).await?;
     // TODO append_shard_doc depends on the initial request, but we don't have access to it
     let mut search_response_rest: ElasticsearchResponse =
-        convert_to_es_search_response(search_response, false, None, None);
+        convert_to_es_search_response(search_response, false, None, None, true)?;
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-serve/src/elasticsearch_api/rest_handler.rs:861]: you are 100% correct

File: quickwit/quickwit-serve/src/elasticsearch_api/rest_handler.rs [hunk 3]
```diff
@@ -941,8 +953,16 @@ fn convert_to_es_search_response(
         },
         aggregations,
         scroll_id: resp.scroll_id,
+        // There is not concept of shards here, but use this to convey split search failures.
+        shards: ShardStatistics {
+            total: 0u32,
+            successful: 0u32,
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-serve/src/elasticsearch_api/rest_handler.rs:959]: That makes sense. I'll do this.

File: quickwit/quickwit-serve/src/elasticsearch_api/rest_handler.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-serve/src/elasticsearch_api/rest_handler.rs]: actually no, `true` was here on purpose, but I added some explanation.

File: quickwit/quickwit-serve/src/elasticsearch_api/rest_handler.rs [hunk 4]
```diff
@@ -852,8 +857,14 @@ async fn es_scroll(
     };
     let search_response: SearchResponse = search_service.scroll(scroll_request).await?;
     // TODO append_shard_doc depends on the initial request, but we don't have access to it
+
+    // Ideally, we would have wanted to reuse the setting from the initial search request.
+    // However, passing that parameter is cumbersome and not necessary:
+    // if we have a valid `scroll_id` it means that the search request was successful.
+    // We can therefore allow failed splits, it won't make any difference.
```
Comment (trinity-1686a) [inline quickwit/quickwit-serve/src/elasticsearch_api/rest_handler.rs:864]: i agree it's cumbersome, and i'm fine with it being left as TODO.
I'm not convinced by the argument that if we have a scroll id, search was successful. After a few pages, we need to do more search, which might hit a missing split (or fail for whatever reason), so an initial success only guarantees we're okay for so much calls.

**Review by trinity-1686a — APPROVED** [review]

File: quickwit/quickwit-serve/src/elasticsearch_api/rest_handler.rs (hunk 4 above)
Comment (fulmicoton) [reply quickwit/quickwit-serve/src/elasticsearch_api/rest_handler.rs:864]: :( yes you are correct.

**Review by trinity-1686a — APPROVED** [review]

**Pull request merged**